### PR TITLE
feat(linter/react): add the `rules_of_hooks` rule.

### PR DESCRIPTION
**Pull request by rzvxa**
[RulesOfHooks](https://github.com/facebook/react/blob/main/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js), [docs](https://react.dev/reference/rules/rules-of-hooks), [npm](https://www.npmjs.com/package/eslint-plugin-react-hooks)
This one sounds like something straight out of `Mortal Kombat`!

### CODE REVIEW

**Comment by rzvxa** [comment]
* **#3071** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3071?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rzvxa and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by rzvxa** [comment]
### Status update: 
This PR passes all react's official tests(even some that aren't passing in the original rule) there is only one situation that isn't detected... The labeled blocks with [labeled] breaks.

These don't have a correct cfg representation.

**Comment by Boshen** [comment]
@mysteryven Good luck reviewing this ... and take your time :-)

**Review by mysteryven — APPROVED** [review]
LGTM, learned something new, thanks!

File: crates/oxc_linter/src/rules/react/rules_of_hooks.rs [hunk 1]
```diff
@@ -0,0 +1,1478 @@
...
+                //      }
+                // }
+                // --------------
+                // function createComponent() {
+                //     return function ComponentWithHookInsideCallback() {
+                //         function handleClick() {
+                //             useState();
+                //         }
+                //     }
+                // }
```
[the remark this answers is not in the document]
Comment (rzvxa) [reply crates/oxc_linter/src/rules/react/rules_of_hooks.rs:218]: Thanks for pointing it out, I'll fix it along with updating the branch.

**Pull request merged**